cpp/src/gandiva/interval_holder.cc

```diff
@@ -0,0 +1,277 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "gandiva/interval_holder.h"
+
+#include "gandiva/node.h"
+#include "gandiva/regex_util.h"
+
+namespace gandiva {
+
+// pre-compiled pattern for matching period that only have numbers
+static const RE2 period_only_contains_numbers(R"(\d+)");
+
+// pre-compiled pattern for matching periods in 8601 formats that not contains weeks.
+static const RE2 iso8601_complete_period(
+    R"(P([[:digit:]]+Y)?([[:digit:]]+M)?([[:digit:]]+D)?)"
+    R"(T([[:digit:]]+H)?([[:digit:]]+M)?([[:digit:]]+S|[[:digit:]]+\.[[:digit:]]+S)?)");
+
+// pre-compiled pattern for matching periods in 8601 formats that not contain time
+// (hours, minutes and seconds) information.
+static const RE2 iso8601_period_without_time(
+    R"(P([[:digit:]]+Y)?([[:digit:]]+M)?([[:digit:]]+D)?)");
+
+// pre-compiled pattern for matching periods in 8601 formats that not contain time
+// (hours, minutes and seconds) information.
+static const std::regex period_not_contains_time(R"(^((?!T).)*$)");
+
+// pre-compiled pattern for matching periods in 8601 formats that contains weeks inside
+// them. The ISO8601 specification defines that if the string contains a week, it can not
+// have other time granularities information, like day, years and months.
+static const RE2 iso8601_period_with_weeks(R"(P([[:digit:]]+W){1})");
+
+int64_t IntervalDaysHolder::GetIntervalDayFromMillis(const std::string& number_as_string,
+                                                     bool* out_valid) {
+  int64_t period_in_millis = std::stol(number_as_string);
+
+  // It considers that a day has exactly 24 hours of duration
+  int64_t days_to_standard_millis = 86400000;
+
+  int64_t qty_days = period_in_millis / days_to_standard_millis;
+  int64_t qty_millis = period_in_millis % days_to_standard_millis;
+
+  // The response is a 64-bit integer where the lower half of the bytes represents the
+  // number of the days and the other half represents the number of milliseconds.
+  int64_t out = (qty_days & 0x00000000FFFFFFFF);
+  out |= ((qty_millis << 32) & 0xFFFFFFFF00000000);
+
+  *out_valid = true;
+  return out;
+}
+
+int64_t IntervalDaysHolder::GetIntervalDayFromWeeks(const std::string& number_as_string,
+                                                    bool* out_valid) {
+  auto qty_weeks = static_cast<int64_t>(std::stod(number_as_string));
```

Comment:
The correct way is to use `stod` in all places because the ISO format allows the user to pass decimal numbers https://en.wikipedia.org/wiki/ISO_8601#Time_intervals.

The reason that I cast as long is that I based on Joda Time implementation: https://www.joda.org/joda-time/apidocs/org/joda/time/Period.html that was the base of Java 8 Period API and that API returns only the integer part of the values.

Reply:
ok. in that case shouldn't you be doing floating point arithmetic and only cast the final qty_millis as double. Otherwise, wouldn't you miss some days or millis.

e,g consider PW30.5

in the above stod returns 30.5. but you cast it to int64_t, which will cause the fractionary part to be lost. Or am I missing something? instead if you use double 
  auto qty_weeks = std::stod(...)
  int64_t millis_in_all_weeks = static_cast<int64_t> (qty_weeks * kMillisInAWeek);
  
you will not loose nearly 1/2 a week.

Reply:
Never mind. I think I understand what you are saying. While time intervals can be specified with decimals and we should accept it, due to the way Joda time works, decimal part should be ignored.
It is still a bit confusing to me, but I think I get it..The only question then is, does the Joda time behaves this way where it accepts the fraction part of the decimal and then simply ignore it or does the Java version return an error if it sees a fraction part..